backend/service/topology/querybuilder_test.go

```diff
@@ -0,0 +1,343 @@
+package topology
+
+import (
+	"testing"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/stretchr/testify/assert"
+
+	topologyv1 "github.com/lyft/clutch/backend/api/topology/v1"
+)
+
+func TestMaxQueryLimit(t *testing.T) {
+	testCases := []struct {
+		id          string
+		input       uint64
+		shouldError bool
+	}{
+		{
+			id:          "Under limit",
+			input:       999,
+			shouldError: false,
+		},
+		{
+			id:          "Equal to limit",
+			input:       1000,
+			shouldError: false,
+		},
+		{
+			id:          "Above limit",
+			input:       1001,
+			shouldError: true,
+		},
+	}
+
+	for _, test := range testCases {
+		_, _, err := paginatedQueryBuilder(&topologyv1.SearchRequest_Filter{}, &topologyv1.SearchRequest_Sort{}, "0", test.input)
+		if test.shouldError {
+			assert.Error(t, err)
+		} else {
+			assert.NoError(t, err)
+		}
+	}
+}
+
+func TestPaginatedQueryBuilder(t *testing.T) {
+	testCases := []struct {
+		id        string
+		filter    *topologyv1.SearchRequest_Filter
+		sort      *topologyv1.SearchRequest_Sort
+		pageToken string
+		limit     uint64
+		expect    string
+	}{
+		{
+			id:        "Default",
+			filter:    &topologyv1.SearchRequest_Filter{},
+			sort:      &topologyv1.SearchRequest_Sort{},
+			pageToken: "0",
+			limit:     0,
+			expect:    "SELECT id, data, metadata FROM topology_cache ORDER BY ID ASC LIMIT 100 OFFSET 0",
+		},
+		{
+			id:        "No page set",
+			filter:    &topologyv1.SearchRequest_Filter{},
+			sort:      &topologyv1.SearchRequest_Sort{},
+			pageToken: "",
+			limit:     5,
+			expect:    "SELECT id, data, metadata FROM topology_cache ORDER BY ID ASC LIMIT 5 OFFSET 0",
+		},
+		{
+			id:        "Page 0 with limit set",
+			filter:    &topologyv1.SearchRequest_Filter{},
+			sort:      &topologyv1.SearchRequest_Sort{},
+			pageToken: "0",
+			limit:     5,
+			expect:    "SELECT id, data, metadata FROM topology_cache ORDER BY ID ASC LIMIT 5 OFFSET 0",
+		},
+		{
+			id:        "Change PageToken and Limits",
+			filter:    &topologyv1.SearchRequest_Filter{},
+			sort:      &topologyv1.SearchRequest_Sort{},
+			pageToken: "10",
+			limit:     5,
+			expect:    "SELECT id, data, metadata FROM topology_cache ORDER BY ID ASC LIMIT 5 OFFSET 50",
+		},
+		{
+			id: "All Options",
+			filter: &topologyv1.SearchRequest_Filter{
+				Search: &topologyv1.SearchRequest_Filter_Search{
+					Field: "metadata.search.field",
+					Text:  "cat",
+				},
+				TypeUrl: "type.googleapis.com/clutch.aws.ec2.v1.AutoscalingGroup",
+				Metadata: map[string]string{
+					"label": "value",
+				},
+			},
+			sort: &topologyv1.SearchRequest_Sort{
+				Field:     "metadata.meow.iam.a.cat",
+				Direction: topologyv1.SearchRequest_Sort_ASCENDING,
+			},
+			pageToken: "10",
+			limit:     5,
+			expect:    "SELECT id, data, metadata FROM topology_cache WHERE metadata->'search'->'field' LIKE $1 AND resolver_type_url = $2 AND metadata @> $3::jsonb ORDER BY $4 ASC LIMIT 5 OFFSET 50",
```

Comment:
have you tested this query against a real DB? i think that it needs to be `metadata->'search'->>'field'` (double `>>`)

Reply:
Ive tested many permutations against a real database yes. I dont think the double `>>` is necessary but I check again.

Reply:
:( yeah `>>` returns text while `>` returns `jsonb`.

Reply:
Actually what am I doing, we dont even support this metadata is just a simple map https://github.com/lyft/clutch/blob/topology-paginated-api-impl/api/topology/v1/topology_api.proto#L226-L228

Reply:
hmm maybe that should be nested like https://github.com/lyft/clutch/blob/topology-paginated-api-impl/api/topology/v1/topology_api.proto#L216

Reply:
Nice let me use that

Reply:
Nice much better.

Given this query
```json
{
    "limit": 100,
    "page_token": "0",
    "sort": {
        "direction": "ASCENDING",
        "field":"column.id"
    },
    "filter": {
        "search": {
            "field":"metadata.levelone.leveltwo.field",
            "text": "value"
        },
        "typeUrl": "type.googleapis.com/clutch.k8s.v1.Pod"
    }
}
```

Produces this SQL 
```sql
SELECT id, data, metadata FROM topology_cache WHERE metadata->'levelone'->'leveltwo'->>'field' LIKE $1 AND resolver_type_url = $2 ORDER BY $3 ASC LIMIT 100 OFFSET 0
```

which yields this result
```json
{
    "resources": [
        {
            "id": "clutch-main-1234",
            "pb": {
                "@type": "type.googleapis.com/clutch.k8s.v1.Pod",
                "cluster": "",
                "namespace": "",
                "name": "",
                "containers": [],
                "nodeIp": "",
                "podIp": "",
                "state": "UNSPECIFIED",
                "startTime": null,
                "labels": {},
                "annotations": {}
            },
            "metadata": {
                "app": "clutch",
                "environment": "staging",
                "levelone": {
                    "leveltwo": {
                        "field": "value"
                    }
                }
            }
        }
    ],
    "nextPageToken": "1"
}